File: linked lists/interviewQNs/reverseALinkedList.py
# Recursion
from os import preadv


class Node:
  def __init__(self,value =0,next =None):
    self.value = value
    self.next = next

class LinkedList:
  def __init__(self):
    self.head = None

# normal recursion
  def reverseLinkedListRecursive(self,head):
      if not head or not head.next:
        return head
      
      reversedTail = self.reverseLinkedListRecursive(head.next)
      head.next.next = head
      head.next = None

      return reversedTail
  
  # Tail recursion
  def tailReverseLinkedListRecursive(self,currentNode,prevNode):
     if currentNode.next is None: # for the last node, make it the head
        self.head = currentNode
        currentNode.next = prevNode
        return
     nextNode = currentNode.next
     currentNode.next = prevNode
     self.tailReverseLinkedListRecursive(nextNode,currentNode)

# Iteration

  def iterativeReverse(self):
        prev = None
        current = self.head
        while current:
            next = current.next
            current.next = prev
            prev = current
            current = next
        self.head = prev


  def reverse(self):
        if self.head is None:
            return
        self.tailReverseLinkedListRecursive(self.head, None)

  def __str__(self):
          linkedListStr = ""
          temp = self.head
          while temp:
              linkedListStr = (linkedListStr +
                              str(temp.value) + " ")
              temp = temp.next
          return linkedListStr
  
  # Pushes new value to the head of the list
  def push(self, value):
      temp = Node(value)
      temp.next = self.head
      self.head = temp

 
linkedList = LinkedList()
linkedList.push(20)
linkedList.push(4)
linkedList.push(15)
linkedList.push(85)

print("Given linked list")
print(linkedList)
 
linkedList.head = linkedList.reverseLinkedListRecursive(linkedList.head)

linkedList.tailReverseLinkedListRecursive(linkedList.head, None)

linkedList.iterativeReverse()
# linkedList.reverse()
print("Reversed linked list")
print(linkedList)

File: linked lists/interviewQNs/reverseALinkedList.c
#include <stdio.h>
#include <stdlib.h>

struct ListNode {
  int value;
  struct ListNode* next; // pointer to the next node
};

struct ListNode* createNode(int value)
{
  struct ListNode* newNode = (struct ListNode*)malloc(sizeof(struct ListNode));
  newNode->value = value;
  newNode->next = NULL;
  return newNode;
};

struct ListNode* reverseListIterative(struct ListNode* head)
{
  struct ListNode* prev = NULL;
  struct ListNode* current = head;

  while (current)
  {
    struct ListNode* nextNode = current->next;
    current->next = prev;
    prev = current;
    current = nextNode;    
  }
  
  return prev;
};

struct ListNode* reverseListRecursive(struct ListNode* head)
{
  if (!head || !head->next)
  {
    return head;
  }

  struct ListNode* reversedTail = reverseListRecursive(head->next);
  head->next->next = head;
  head->next = NULL;

  return reversedTail;
};

struct ListNode* reverseListTailRecursive(struct ListNode* head, struct ListNode* prev)
{
  if (!head)
  {
    return prev;
  }

  struct ListNode* nextNode = head->next;
  head->next = prev;

  return reverseListTailRecursive(nextNode, head);
};


int main() {
    // Create a sample linked list: 1 -> 2 -> 3 -> 4
    struct ListNode* head = createNode(1);
    head->next = createNode(2);
    head->next->next = createNode(3);
    head->next->next->next = createNode(4);

    // Print the original linked list
    printf("Original Linked List: ");
    struct ListNode* current = head;
    while (current) {
        printf("%d -> ", current->value);
        current = current->next;
    }
    printf("NULL\n");

    // Reverse the linked list iteratively
    head = reverseListIterative(head);

    // Print the reversed linked list
    printf("Reversed Linked List (Iterative): ");
    current = head;
    while (current) {
        printf("%d -> ", current->value);
        current = current->next;
    }
    printf("NULL\n");

    // Reverse the linked list recursively
    head = reverseListRecursive(head);

    // Print the reversed linked list
    printf("Reversed Linked List (Recursive): ");
    current = head;
    while (current) {
        printf("%d -> ", current->value);
        current = current->next;
    }
    printf("NULL\n");


    // tail recursion
    head = reverseListTailRecursive(head,NULL);
    printf("Tail recursive reversed Linked List: ");
    current = head;
    while (current) 
    {
      printf("%d -> ",current->value);
      current = current->next;
    }
    printf("NULL\n");

    return 0;
}
